[Swift] Get list of DynamicObject

tom-sparo opened the pull request:
Hey @jpsim -- this pull request implements [#2685](https://github.com/realm/realm-cocoa/issues/2685)

Wasn't sure what style you all wanted for documentation/testing so I just guestimated based on code elsewhere (sorta shooting from the hip!)

As a user I can already get this functionality myself by using an unsafeBitCast on the result of KVC, but it's nice to wrap away these types of casts so the user doesn't feel like they're doing something wrong.

Let me know what you think!

jpsim commented:
Thanks for the PR :sparkles:! It always feels awesome to see users contributing. Here are a few points of feedback:

1. All Swift changes should generally be applied to both Swift 1.2 and Swift 2.0 (there are two versioned RealmSwift directories in the root of the repo).
2. Tests should access members of the `List` rather than just check the empty case, since that's not comprehensive enough.

Thanks again.

tom-sparo commented:
Added in change to the Swift 1.2 branch and added in more robust testing -- let me know if you want more than that!

Also can't see what the failing checks are so if there are compile errors or test errors let me know so I can fix them.  I assumed there's no way for me to compile the repo to check myself because Core isn't open source, but if there's a way to compile without it I missed let me know so I can save you reading time.

jpsim commented:
We use an internal Jenkins setup for our CI, so we unfortunately can't easily share that for external contributors :cry:. This is something I'd love to address, but simply don't have the time.

But you can run the same gamut of tests we do by running `sh build.sh verify` assuming you have Xcode 6.4 (with iOS 7 simulators) and Xcode 7.0.1 installed on your system.

The whole set of tests take a really long time to run (we parallelize them on ~12 CI workers!), but you can usually get a good sense of whether or not things are broken by running the following small(er) tests:

```
./build.sh verify-osx
export REALM_SWIFT_VERSION=1.2
./build.sh verify-osx-swift
export REALM_SWIFT_VERSION=2.0
./build.sh verify-osx-swift
```

In your case, the Swift 1.2 tests are failing:

```
Testing failed:
	Consecutive statements on a line must be separated by ';'
	Consecutive statements on a line must be separated by ';'
	Use of unresolved identifier 'try'
	'List<SwiftStringObject>' does not have a member named 'appendContentsOf'
	Use of unresolved identifier 'try'
```

tom-sparo commented:
Ah great!  That's at least better than no testing at all.  Both my local tests (for swift 1.2 and swift 2.0) passed so I'll have to rely on you all again to paste errors :disappointed:

jpsim commented on RealmSwift-swift1.2/Object.swift:
```diff
@@ -191,6 +191,27 @@ public class Object: RLMObjectBase, Equatable, Printable {
+    :nodoc:
+    */
+    public func dynamicList(propertyName: String) -> List<DynamicObject> {
+        return unsafeBitCast(self.listForProperty(RLMValidatedGetProperty(self, propertyName)), List<DynamicObject>.self)
```
`self.` can probably be omitted here.

jpsim commented on RealmSwift-swift1.2/Tests/ObjectTests.swift:
```diff
@@ -284,4 +284,22 @@ class ObjectTests: TestCase {
+        let arrayObject = SwiftArrayPropertyObject()
+        let str1 = SwiftStringObject()
+        let str2 = SwiftStringObject()
+        arrayObject.array.append(str1)
```
`arrayObject.array.extend([str1, str2])`?

jpsim commented on RealmSwift-swift1.2/Object.swift:
```diff
@@ -191,6 +191,27 @@ public class Object: RLMObjectBase, Equatable, Printable {
+    // MARK: Dynamic list
+
+    /**
+    This method is useful only in specialized circumstances, for example, when building
```
This comment is written in the Swift 2.0 documentation style. Please use `:warning:`, `:param: propertyName` and `:returns:`.

jpsim commented:
Thanks for the PR fixes, @tom-sparo! I think this could be merged after you address my last round of (nitpicky) feedback.

Our CI jobs are seeing some device test failures at the moment because we're adding some additional iOS devices and still working out some kinks, so we may merged this despite GitHub showing a red 'x'.

tom-sparo commented:
Fixed all feedback!

jpsim commented:
This looks all good to me, how about for you @bigfish24?

bigfish24 commented:
:+1: thanks for the contribution @tom-sparo!
